DB/src/main/java/io/deephaven/db/v2/by/StreamFirstByChunkedOperator.java

```diff
@@ -0,0 +1,237 @@
+package io.deephaven.db.v2.by;
+
+import io.deephaven.base.verify.Assert;
+import io.deephaven.db.tables.Table;
+import io.deephaven.db.tables.select.MatchPair;
+import io.deephaven.db.v2.QueryTable;
+import io.deephaven.db.v2.ShiftAwareListener;
+import io.deephaven.db.v2.sources.*;
+import io.deephaven.db.v2.sources.chunk.Attributes.*;
+import io.deephaven.db.v2.sources.chunk.*;
+import io.deephaven.db.v2.utils.Index;
+import io.deephaven.db.v2.utils.OrderedKeys;
+import io.deephaven.db.v2.utils.ReadOnlyIndex;
+import io.deephaven.db.v2.utils.ShiftedOrderedKeys;
+import io.deephaven.util.SafeCloseableList;
+import org.jetbrains.annotations.NotNull;
+
+import java.lang.ref.SoftReference;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * A firstBy aggregation operator for stream tables.
+ *
+ * @see Table#STREAM_TABLE_ATTRIBUTE
+ */
+public class StreamFirstByChunkedOperator implements IterativeChunkedAggregationOperator {
+
+    private static final int COPY_CHUNK_SIZE = ArrayBackedColumnSource.BLOCK_SIZE;
+
+    private final int numResultColumns;
+    private final ColumnSource<?>[] inputColumns;
+    private final WritableSource<?>[] outputColumns;
+
+    private final Map<String, ArrayBackedColumnSource<?>> resultColumns;
+
+    private long previousTableSize;
+    private SoftReference<LongArraySource> cachedRedirections;
+    private LongArraySource redirections;
+
+    StreamFirstByChunkedOperator(@NotNull final MatchPair[] resultPairs, @NotNull final Table streamTable) {
+        numResultColumns = resultPairs.length;
+        inputColumns = new ColumnSource[numResultColumns];
+        outputColumns = new WritableSource[numResultColumns];
+        final Map<String, ArrayBackedColumnSource<?>> resultColumnsMutable = new LinkedHashMap<>(numResultColumns);
+        for (int ci = 0; ci < numResultColumns; ++ci) {
+            final MatchPair resultPair = resultPairs[ci];
+            final ColumnSource<?> streamSource = streamTable.getColumnSource(resultPair.left());
+            final ArrayBackedColumnSource<?> resultSource = ArrayBackedColumnSource.getMemoryColumnSource(0, streamSource.getType(), streamSource.getComponentType());
+            resultColumnsMutable.put(resultPair.left(), resultSource);
+            inputColumns[ci] = ReinterpretUtilities.maybeConvertToPrimitive(streamSource);
+            outputColumns[ci] = (WritableSource<?>) ReinterpretUtilities.maybeConvertToPrimitive(resultSource);
```

Comment:
This feels backwards; that we would want to have the outputColumn be an array backed column source of the primitive not of the type from the streamSource; and then you would wrap it in the opposite.

Reply:
I'm pretty sure you're wrong about this; we (you?) did a good job designing the array-backed column sources for this use case:
`DateTimeArraySource.ReinterpretedAsLong` is just as efficient for filling with longs as `LongArraySource`, and `DateTimeArraySource` is more efficient for getting `DBDateTimes` out of than `LongAsDateTimeColumnSource` wrapped around `LongArraySource`.
BooleanArraySource.ReinterpretedAsByte` is similarly good for filling with bytes, and I think `BooleanArraySource` is way better for reading `Booleans` from than `ByteAsBooleanColumnSource` wrapped around `ByteArraySource`.